codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java

```diff
@@ -253,6 +253,17 @@
     }
 
     /**
+     * A key to be used for associating application-defined properties with streams within this connection.
+     */
+    interface PropertyKey {
+    }
```

Comment:
Not providing the index is a way of hiding the implementation details.  I don't see why you consider the cast dangerous.  If the wrong key type were provided at runtime it would generate a `ClassCastException`, which seems perfectly reasonable to me.  Would you prefer we throw `IllegalArgumentException` instead?  I guess I'm not quite sure I understand your concern.

Regarding the question of ensuring the key came from the appropriate connection instance, I could add this to the key if you think it's useful.